### PR TITLE
Implement support for update pages API

### PR DESCRIPTION
**Pull request by patrickmarabeas**
Implements support for updating Github Pages sites - allowing CName and
Source to be customized.

Fixes: #1361

### CODE REVIEW

File: github/repos_pages.go [hunk 1]
```diff
@@ -75,6 +75,34 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
 	return enable, resp, nil
 }
 
+// PagesUpdate sets up parameters needed to update a GitHub Pages site.
+type PagesUpdate struct {
+	CNAME  *string `json:"cname,omitempty"`
+	Source *string `json:"source,omitempty"`
```
Comment (gmlewis) [inline github/repos_pages.go]: Can you please add a comment above `Source` like this:

```
// Source must include the branch name, and may optionally specify the subdirectory "/docs".
// Possible values are: "gh-pages", "master", and "master /docs".
```

File: github/repos_pages.go [hunk 2]
```diff
@@ -75,6 +75,34 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
 	return enable, resp, nil
 }
 
+// PagesUpdate sets up parameters needed to update a GitHub Pages site.
+type PagesUpdate struct {
+	CNAME  *string `json:"cname,omitempty"`
```
Comment (gmlewis) [inline github/repos_pages.go]: Can you please add a comment above `CNAME` like this:

```
// CNAME represents a custom domain for the repository.
// Leaving CNAME empty will remove the custom domain.
```

File: github/repos_pages.go [hunk 3]
```diff
@@ -75,6 +75,34 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
...
+// GitHub API docs: https://developer.github.com/v3/repos/pages/#update-information-about-a-pages-site
+func (s *RepositoriesService) UpdatePages(ctx context.Context, owner, repo string, opts *PagesUpdate) (*Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/pages", owner, repo)
+
+	req, err := s.client.NewRequest("PUT", u, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", mediaTypeEnablePagesAPIPreview)
```
Comment (gmlewis) [inline github/repos_pages.go]: If I'm reading the GitHub API docs correctly, this custom preview header is not needed for this endpoint, so let's please remove it.

File: github/repos_pages.go [hunk 4]
```diff
@@ -75,6 +75,34 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
...
+
+	update := new(PagesUpdate)
```
Comment (gmlewis) [inline github/repos_pages.go]: No return value is desired, so please remove this line.

File: github/repos_pages.go [hunk 5]
```diff
@@ -75,6 +75,34 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
...
+	resp, err := s.client.Do(ctx, req, update)
```
Comment (gmlewis) [inline github/repos_pages.go]: s/update/nil/

File: github/repos_pages_test.go [hunk 6]
```diff
@@ -52,6 +52,35 @@ func TestRepositoriesService_EnablePages(t *testing.T) {
...
+		CNAME:  String("www.my-domain.com"),
+		Source: String("gh-pages"),
+	}
+
+	mux.HandleFunc("/repos/o/r/pages", func(w http.ResponseWriter, r *http.Request) {
+		v := new(PagesUpdate)
+		json.NewDecoder(r.Body).Decode(v)
+
+		testMethod(t, r, "PUT")
+		testHeader(t, r, "Accept", mediaTypeEnablePagesAPIPreview)
```
Comment (gmlewis) [inline github/repos_pages_test.go]: This line can be removed.

File: github/repos_pages.go (hunk 2 above)
Comment (patrickmarabeas) [reply github/repos_pages.go]: Sorry, wasn't expecting such a prompt response, was still poking at it.

Thoughts on the secondary type I've since added so this can remain optional without removing the custom domain?

File: github/repos_pages.go [hunk 7]
```diff
@@ -75,6 +75,53 @@ func (s *RepositoriesService) EnablePages(ctx context.Context, owner, repo strin
...
+// GitHub API docs: https://developer.github.com/v3/repos/pages/#update-information-about-a-pages-site
+func (s *RepositoriesService) UpdatePages(ctx context.Context, owner, repo string, opts *PagesUpdate) (*Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/pages", owner, repo)
+
+	req, err := s.client.NewRequest("PUT", u, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if *opts.CNAME == "" {
```
Comment (gmlewis) [inline github/repos_pages.go]: Let's delete this block completely (lines 107-116) and remove the `pagesUpdateNullCNAME` and then simply remove the `,omitempty` on line 82.

That way, if `CNAME` is defined, it is used, but if `CNAME` is empty, `null` will be sent and the custom domain will be removed.

File: github/repos_pages_test.go [hunk 8]
```diff
@@ -52,6 +52,34 @@ func TestRepositoriesService_EnablePages(t *testing.T) {
 	}
 }
 
+func TestRepositoriesService_UpdatePages(t *testing.T) {
```
Comment (gmlewis) [inline github/repos_pages_test.go]: You could add a second test proving that when `CNAME` is empty, `null` is truly passed.
An example like that is `TestGitService_CreateTree_Delete` in `git_trees_test.go`.

File: github/repos_pages.go (hunk 7 above)
Comment (patrickmarabeas) [reply github/repos_pages.go]: Sure.

File: github/repos_pages_test.go (hunk 8 above)
Comment (patrickmarabeas) [reply github/repos_pages_test.go]: Added test.

File: github/repos_pages_test.go [hunk 9]
```diff
@@ -52,6 +52,61 @@ func TestRepositoriesService_EnablePages(t *testing.T) {
...
+		Source: String("gh-pages"),
+	}
+
+	mux.HandleFunc("/repos/o/r/pages", func(w http.ResponseWriter, r *http.Request) {
+		v := new(PagesUpdate)
+		json.NewDecoder(r.Body).Decode(v)
+
+		testMethod(t, r, "PUT")
+		want := &PagesUpdate{CNAME: nil, Source: String("gh-pages")}
+		if !reflect.DeepEqual(v, want) {
```
Comment (gmlewis) [inline github/repos_pages_test.go]: Unfortunately, this doesn't test that a `null` was actually sent via JSON.

Here's the example I was talking about:

```
		got, err := ioutil.ReadAll(r.Body)
		if err != nil {
			t.Fatalf("unable to read body: %v", err)
		}

		testMethod(t, r, "POST")

		want := []byte(`{"base_tree":"b","tree":[{"sha":null,"path":"content.md","mode":"100644"}]}` + "\n")
		if !bytes.Equal(got, want) {
			t.Errorf("Git.CreateTree request body: %s, want %s", got, want)
		}
```

Comment (patrickmarabeas) [reply github/repos_pages_test.go]: Ah, right.

**Comment by gmlewis** [comment]
Thank you, @wesleimp! Merging.
